File: MATLAB_models/solar_class.m
classdef solar_class
    properties
        location
        mirror_type
        nominal_DNI
        solar_multiple
        Tmax
        Tmin

        power
        dumped
        DNI
        nY % number of years

        land_mult
        
        mirror_cost_unit
        land_cost_unit
        HX_cost_unit

        mirror_cost
        land_cost
        HX_cost
        total_cost

        nominal_eff
        annual_eff
        mirror_aperture
        land_area

        available_solar_heat % GWh-th
        solar_thermal_generated % GWh-th
        total_dumped % GWh-th

    end

    methods
        function obj = solar_class(location, mirror_type, nominal_DNI, solar_multiple, Tmax, Tmin, land_mult, CSP_cost, nY)
            obj.location = location ;
            obj.mirror_type = mirror_type ;
            obj.nominal_DNI = nominal_DNI ;
            obj.solar_multiple = solar_multiple ;
            obj.Tmax = Tmax ;
            obj.Tmin = Tmin ;

            obj.land_mult = land_mult ;
            obj.nY = nY ;

            obj.DNI = zeros(8760,1);
            obj.power = zeros(nY*8760,1);
            obj.dumped = zeros(nY*8760,1);
            
            obj.mirror_cost_unit = CSP_cost.mirror ;
            obj.land_cost_unit = CSP_cost.land;
            obj.HX_cost_unit = CSP_cost.HX ;

        end


        function obj = CSP_annual_energy(obj)
            n = (obj.nY-1)*8760+1:obj.nY*8760;

            obj.available_solar_heat = sum(obj.DNI) * obj.mirror_aperture / 1e9 ;
            obj.solar_thermal_generated = sum(obj.power(n)) / 1000;
            obj.total_dumped = sum(obj.dumped(n)) / 1000 ;
            
        end

        % Calculate the solar field cost
        function obj = calc_CSP_cost(obj,PC)

            obj.mirror_cost = obj.mirror_cost_unit * obj.mirror_aperture ;
            obj.land_cost   = obj.land_cost_unit   * obj.land_area ;

            obj.HX_cost     = obj.HX_cost_unit * PC.Qin0 * 1000 ;

            obj.total_cost = obj.mirror_cost + obj.land_cost + obj.HX_cost ;

        end

    end


end

File: Python_TEA_model/solar_class.py
import numpy as np


class solar_class:
    def __init__(
        self,
        location,
        mirror_type,
        nominal_DNI,
        solar_multiple,
        initial_size,
        Tmax,
        Tmin,
        land_mult,
        CSP_cost,
        nY=2,
    ):
        self.location = location
        self.mirror_type = mirror_type
        self.nominal_DNI = float(nominal_DNI)
        self.solar_multiple = float(solar_multiple)
        self.initial_size = float(initial_size)
        self.Tmax = float(Tmax)
        self.Tmin = float(Tmin)

        self.land_mult = float(land_mult)
        self.nY = int(nY)

        self.DNI = np.zeros(8760, dtype=float)
        self.power = np.zeros(self.nY * 8760, dtype=float)
        self.dumped = np.zeros(self.nY * 8760, dtype=float)

        self.mirror_cost_unit = float(CSP_cost["mirror"])
        self.land_cost_unit = float(CSP_cost["land"])
        self.HX_cost_unit = float(CSP_cost["HX"])

        self.mirror_cost = 0.0
        self.land_cost = 0.0
        self.HX_cost = 0.0
        self.total_cost = 0.0

        self.nominal_eff = 0.0
        self.annual_eff = 0.0
        self.mirror_aperture = 0.0
        self.land_area = 0.0

        self.available_solar_heat = 0.0
        self.solar_thermal_generated = 0.0
        self.total_dumped = 0.0

    def CSP_annual_energy(self):
        hours = 8760
        hours2 = self.nY * hours
        start = (self.nY - 1) * hours

        self.available_solar_heat = float(np.sum(self.DNI) * self.mirror_aperture / 1e9)
        self.solar_thermal_generated = float(np.sum(self.power[start:hours2]) / 1000.0)
        self.total_dumped = float(np.sum(self.dumped[start:hours2]) / 1000.0)

        return self

    def calc_CSP_cost(self, PC):
        self.mirror_cost = self.mirror_cost_unit * self.mirror_aperture
        self.land_cost = self.land_cost_unit * self.land_area
        self.HX_cost = self.HX_cost_unit * PC.Qin0 * 1000.0
        self.total_cost = self.mirror_cost + self.land_cost + self.HX_cost
        return self
